miniapps/shifted/integ_algoim.hpp

```diff
@@ -0,0 +1,351 @@
+{
+   if (p == 0)
+   {
+      u[0] = float_type(1.);
+   }
+   else
+   {
+      int i;
+      const int *b = Poly_1D::Binom(p);
+      float_type z = x;
+      for (i = 1; i < p; i++)
+      {
+         u[i] = b[i]*z;
+         z *= x;
+      }
+      u[p] = z;
+      z = y;
+      for (i--; i > 0; i--)
+      {
+         u[i] *= z;
+         z *= y;
+      }
+      u[0] = z;
+   }
+}
+
+/// Templated version of CalcBinomTerms
+template<typename float_type>
+void CalcBinomTerms(const int p, const float_type x, const float_type y,
+                    float_type* u, float_type* d)
+{
+   if (p == 0)
+   {
+      u[0] = float_type(1.);
+      d[0] = float_type(0.);
+   }
+   else
+   {
+      int i;
+      const int *b = Poly_1D::Binom(p);
+      const float_type xpy = x + y, ptx = p*x;
+      float_type z = float_type(1.);
+
+      for (i = 1; i < p; i++)
+      {
+         d[i] = b[i]*z*(i*xpy - ptx);
+         z *= x;
+         u[i] = b[i]*z;
+      }
+      d[p] = p*z;
+      u[p] = z*x;
+      z = float_type(1.);
+      for (i--; i > 0; i--)
+      {
+         d[i] *= z;
+         z *= y;
+         u[i] *= z;
+      }
+      d[0] = -p*z;
+      u[0] = z*y;
+   }
+
+}
+
+/// Templated version of CalcDBinomTerms
+template <typename float_type>
+void CalcDBinomTerms(const int p, const float_type x,
+                     const float_type y, float_type* d)
+{
+   if (p == 0)
+   {
+      d[0] = 0.;
+   }
+   else
+   {
+      int i;
+      const int *b = Poly_1D::Binom(p);
+      const float_type xpy = x + y, ptx = p*x;
+      float_type z = float_type(1.);
+
+      for (i = 1; i < p; i++)
+      {
+         d[i] = b[i]*z*(i*xpy - ptx);
+         z *= x;
+      }
+      d[p] = p*z;
+      z = float_type(1.);
+      for (i--; i > 0; i--)
+      {
+         d[i] *= z;
+         z *= y;
+      }
+      d[0] = -p*z;
+   }
+}
+
+/// Templated evaluation of Bernstein basis
+template <typename float_type>
+void CalcBernstein(const int p, const float_type x, float_type *u)
+{
+   CalcBinomTerms(p, x, 1. - x, u);
+}
+
+
+/// Templated evaluation of Bernstein basis
+template <typename float_type>
+void CalcBernstein(const int p, const float_type x,
+                   float_type *u, float_type *d)
+{
+   CalcBinomTerms(p, x, 1. - x, u, d);
+}
+
+
+}
+
+/// Construct volumetric and surface integration rules for a given element
+/// using the Algoim library. The volume is define as the possitive part of
+/// a level-set function(LSF) (lsfun argument in the constructor). The surface
+/// is defined as the zero level-set of the LSF.
+class AlgoimIntegrationRule
+{
+public:
+
+   /// Construct Algoim object using a finite element, its transformation
+   /// and level-set function defined over the element using Lagrangian
+   /// bases. The argument o provides the number of the integration points
+   /// of the 1D Gaussian integration used for deriving the vol/surface
+   /// integration rules.
+   AlgoimIntegrationRule(int o, const FiniteElement &el,
```

Comment:
Can we move the implementation of the constructor to the .cpp file?

Reply:
moved